packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -112,6 +119,121 @@ exports.sourceNodes = async (
     pluginConfig,
   })
 
+  createTypes(`
+  interface ContentfulEntry @nodeInterface {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(`
+  interface ContentfulReference {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(
+    schema.buildObjectType({
+      name: `ContentfulAsset`,
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `Node`],
+    })
+  )
+
+  const gqlTypes = contentTypeItems.map(contentTypeItem =>
+    schema.buildObjectType({
+      name: _.upperFirst(
+        _.camelCase(
+          `Contentful ${
+            pluginConfig.get(`useNameForId`)
+              ? contentTypeItem.name
+              : contentTypeItem.sys.id
+          }`
+        )
+      ),
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `ContentfulEntry`, `Node`],
+    })
+  )
+  createTypes(gqlTypes)
+
+  console.time(`Process Contentful data`)
+
+  // Remove deleted entries and assets from cached sync data set
+  previousSyncData.entries = previousSyncData.entries.filter(
+    entry =>
+      _.findIndex(
+        currentSyncData.deletedEntries,
+        o => o.sys.id === entry.sys.id
+      ) !== -1
+  )
+
+  previousSyncData.assets = previousSyncData.assets.filter(
+    asset =>
+      _.findIndex(
+        currentSyncData.deletedAssets,
+        o => o.sys.id === asset.sys.id
+      ) !== -1
+  )
+
+  // Merge cached data with new sync data
+  // order is important here, fresh data first
+  currentSyncData.entries = _.uniqBy(
```

Comment:
I imagine `uniqBy` has non-linear perf implications as well. Probably much faster with the aforementioned pre-generated `Map` cache.